frontend/src/app/reports/reports.component.ts

```diff
@@ -0,0 +1,301 @@
+
+            const year = key.substr( key.length - 5, 4);
+            if (categories.indexOf(year) === -1) {
+                categories.push(year);
+            }
+        }
+        this.groupedBarPlot(groupedColumnsData, categories, true);
+    }
+
+    /**
+     * converts json for number of observations by week to plot data
+     */
+    makeBarPlotNumObsByWeek(data: Object) {
+        // todo: remove demo data
+        // const responseByWeek = {
+        //     '1st (2017)': 100,
+        //     '2nd (2017)': 200,
+        //     '3rd (2017)': 300,
+        //     '4th (2017)': 400,
+        //     '1st (2018)': 500,
+        //     '4th (2018)': 600
+        // };
+
+        const categories: string[] = new Array();
+        const groupedColumnsData: [[string | number]] = [['1']];
+        const referenceWeeks: string[] = ['1st'];
+
+        for (let i = 2; i <= 52; i++) {
+            groupedColumnsData.push(['' + i]);
+            referenceWeeks.push(this.getGetOrdinal(i));
+        }
+
+        for (const key of Object.keys(data)) {
+            const endOfWeekOrdName = key.indexOf(' (');
+            const week = key.substr(0, endOfWeekOrdName);
+            groupedColumnsData[referenceWeeks.indexOf(week)].push(data[key]);
+
+            const year = key.substr( key.length - 5, 4);
+            if (categories.indexOf(year) === -1) {
+                categories.push(year);
+            }
+        }
+        this.groupedBarPlot(groupedColumnsData, categories, true);
+    }
+
+    /**
+     * a helper function to add bar plot data to this.chart
+     */
+    groupedBarPlot(groupedColumnsData, categories, legendShow) {
+        // todo: fully implement pf style
+        // const verticalBarChartConfig = patternfly.c3ChartDefaults().getDefaultGroupedBarConfig(categories);
+        // verticalBarChartConfig.bindto = '#chart';
+        // verticalBarChartConfig.data = {
+        //             columns: groupedColumnsData,
+        //             type: 'bar'};
+        // verticalBarChartConfig.axis = {
+        //             x: {categories: categories,
+        //                 type: 'category'}};
+        // verticalBarChartConfig.legend = {show: legendShow};
+        // this.chart = c3.generate(verticalBarChartConfig);
+        this.chart = c3.generate({
+            bindto: '#chart',
+            data: {
+                columns: groupedColumnsData,
+                type: 'bar'
+            },
+            axis: {
+                x: {
+                    categories: categories,
+                    type: 'category'
+                }
+            },
+            legend: {
+                show: legendShow
+            }
+        });
+    }
+
+    /**
+     * a helder function to convert a number to its ordinal (1st, 2nd, etc.) string value
+     */
+    getGetOrdinal(n: number): string {
+        const s = ['th', 'st', 'nd', 'rd'],
+            v = n % 100;
+        return n + (s[(v - 20) % 10] || s[v] || s[0]);
+    }
```

Comment:
This looks like black magic. Maybe renaming some variables would be helpful. But if it works it works.